ansible: Install AVA as a systemd service

moreati opened the pull request:
This playbook

 - Installs Gecko dependencies (Ubuntu 20.04 only for now)
 - Clones & builds ava-labs/gecko
 - Creates an ava user
 - Installs Gecko in `/usr/bin`
 - Creates and installs a staking certificate
 - Installs Gecko as a systemd service called "ava"
 - Configures `/var/lib/ava/db` as the database
 - Configures `/var/log/ava` as the log destination
 - Starts the service

<details>
<summary>An example run (click to expand)</summary>

```
$ ansible-playbook -i inventory.yml service_playbook.yml

PLAY [Configure AVA service] ********************************************************************************

TASK [Gathering Facts] **************************************************************************************
ok: [droplet]

TASK [ava-base : Install deps] ******************************************************************************
ok: [droplet]

TASK [gopath : Set GOPATH] **********************************************************************************
ok: [droplet]

TASK [ava-build : Update git clone] *************************************************************************
ok: [droplet]

TASK [ava-build : Build project] ****************************************************************************
changed: [droplet]

TASK [ava-certs : Create staker key] ************************************************************************
ok: [droplet]

TASK [ava-certs : Create staker certificate request] ********************************************************
ok: [droplet]

TASK [ava-certs : Create staker certificate] ****************************************************************
ok: [droplet]

TASK [ava-user : Create AVA daemon group] *******************************************************************
ok: [droplet]

TASK [ava-user : Create AVA daemon user] ********************************************************************
[WARNING]: The value False (type bool) in a string field was converted to 'False' (type string). If this
does not look like what you expect, quote the entire value to ensure it does not change.
ok: [droplet]

TASK [ava-install : Create directories] *********************************************************************
ok: [droplet] => (item=/var/lib/ava)
ok: [droplet] => (item=/var/lib/ava/keys)
ok: [droplet] => (item=/var/log/ava)
ok: [droplet] => (item=/usr/lib/ava/plugins)

TASK [ava-install : Install binary] *************************************************************************
ok: [droplet]

TASK [ava-install : Install plugins] ************************************************************************
ok: [droplet] => (item={'path': '~/go/src/github.com/ava-labs/gecko/build/plugins/evm'})

TASK [ava-install : Install staking files] ******************************************************************
ok: [droplet] => (item={'src': '~/go/src/github.com/ava-labs/gecko/keys/staker.key', 'mode': 'u=r,go='})
ok: [droplet] => (item={'src': '~/go/src/github.com/ava-labs/gecko/keys/staker.crt', 'mode': 'ugo=r'})

TASK [ava-service : Configure AVA service] ******************************************************************
ok: [droplet]

TASK [ava-service : Enable AVA service] *********************************************************************
ok: [droplet]

PLAY RECAP **************************************************************************************************
droplet                    : ok=16   changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

</details>

### TODO

- [x] Install under `/usr/local`

moreati commented:
```
# systemctl status ava -l | cat
● ava.service - AVA test node
     Loaded: loaded (/etc/systemd/system/ava.service; enabled; vendor preset: enabled)
     Active: active (running) since Sat 2020-05-09 21:15:58 UTC; 57min ago
       Docs: https://docs.ava.network/
   Main PID: 121344 (ava)
      Tasks: 18 (limit: 1137)
     Memory: 87.8M
     CGroup: /system.slice/ava.service
             ├─121344 /usr/bin/ava --public-ip=178.62.34.87 --db-dir=/var/lib/ava/db --plugin-dir=/usr/lib/ava/plugins --log-dir=/var/log/ava --log-level=info
             ├─121407 sh -c /usr/lib/ava/plugins/evm
             └─121408 /usr/lib/ava/plugins/evm

May 09 22:12:22 ubuntu-s-1vcpu-1gb-lon1-01 ava[121344]: 2020-05-09 22:12:22.261200 [net info] a28556eb8b: ping from <Conn fd=50 addr=<NetAddr 18.234.153.22:21001> mode=active>
May 09 22:12:25 ubuntu-s-1vcpu-1gb-lon1-01 ava[121344]: 2020-05-09 22:12:25.920303 [net info] a28556eb8b: ping from <Conn fd=52 addr=<NetAddr 107.23.241.199:21001> mode=active>
May 09 22:12:31 ubuntu-s-1vcpu-1gb-lon1-01 ava[121344]: 2020-05-09 22:12:31.059780 [net info] a28556eb8b: ping from <Conn fd=54 addr=<NetAddr 54.197.215.186:21001> mode=active>
May 09 22:12:33 ubuntu-s-1vcpu-1gb-lon1-01 ava[121344]: 2020-05-09 22:12:33.707433 [net info] a28556eb8b: ping from <Conn fd=48 addr=<NetAddr 34.207.133.167:21001> mode=active>
May 09 22:12:40 ubuntu-s-1vcpu-1gb-lon1-01 ava[121344]: 2020-05-09 22:12:40.002157 [net info] a28556eb8b: ping from <Conn fd=46 addr=<NetAddr 3.227.207.132:21001> mode=active>
May 09 22:13:05 ubuntu-s-1vcpu-1gb-lon1-01 ava[121344]: 2020-05-09 22:13:05.259493 [net info] a28556eb8b: ping from <Conn fd=50 addr=<NetAddr 18.234.153.22:21001> mode=active>
May 09 22:13:06 ubuntu-s-1vcpu-1gb-lon1-01 ava[121344]: 2020-05-09 22:13:06.699653 [net info] a28556eb8b: ping from <Conn fd=52 addr=<NetAddr 107.23.241.199:21001> mode=active>
May 09 22:13:07 ubuntu-s-1vcpu-1gb-lon1-01 ava[121344]: 2020-05-09 22:13:07.105992 [net info] a28556eb8b: ping from <Conn fd=54 addr=<NetAddr 54.197.215.186:21001> mode=active>
May 09 22:13:08 ubuntu-s-1vcpu-1gb-lon1-01 ava[121344]: 2020-05-09 22:13:08.807134 [net info] a28556eb8b: ping from <Conn fd=48 addr=<NetAddr 34.207.133.167:21001> mode=active>
May 09 22:13:22 ubuntu-s-1vcpu-1gb-lon1-01 ava[121344]: 2020-05-09 22:13:22.377751 [net info] a28556eb8b: ping from <Conn fd=46 addr=<NetAddr 3.227.207.132:21001> mode=active>
```

moreati commented:
```
# find /var/lib/ava/
/var/lib/ava/
/var/lib/ava/keys
/var/lib/ava/keys/staker.crt
/var/lib/ava/keys/staker.key
/var/lib/ava/db
/var/lib/ava/db/cascade
/var/lib/ava/db/cascade/v0.2.0
/var/lib/ava/db/cascade/v0.2.0/000022.ldb
/var/lib/ava/db/cascade/v0.2.0/LOG
/var/lib/ava/db/cascade/v0.2.0/MANIFEST-000021
/var/lib/ava/db/cascade/v0.2.0/CURRENT
/var/lib/ava/db/cascade/v0.2.0/CURRENT.bak
/var/lib/ava/db/cascade/v0.2.0/000023.ldb
/var/lib/ava/db/cascade/v0.2.0/000019.ldb
/var/lib/ava/db/cascade/v0.2.0/000020.log
/var/lib/ava/db/cascade/v0.2.0/LOCK
```
```
# find /var/log/ava/
/var/log/ava/
/var/log/ava/node
/var/log/ava/node/0.log
/var/log/ava/node/chain
/var/log/ava/node/chain/pUASyygGd8XyCgzeP9KC1GAiZ4Qjycq6fs2HfphA6vSFbvLe6
/var/log/ava/node/chain/pUASyygGd8XyCgzeP9KC1GAiZ4Qjycq6fs2HfphA6vSFbvLe6/0.log
/var/log/ava/node/chain/pUASyygGd8XyCgzeP9KC1GAiZ4Qjycq6fs2HfphA6vSFbvLe6/http
/var/log/ava/node/chain/pUASyygGd8XyCgzeP9KC1GAiZ4Qjycq6fs2HfphA6vSFbvLe6/http/0.log
/var/log/ava/node/chain/2kKSkL4vgVR6QSknoKprsGHVHiEbf7WiGEeSuJUGoB3WzeBhce
/var/log/ava/node/chain/2kKSkL4vgVR6QSknoKprsGHVHiEbf7WiGEeSuJUGoB3WzeBhce/0.log
/var/log/ava/node/chain/2kKSkL4vgVR6QSknoKprsGHVHiEbf7WiGEeSuJUGoB3WzeBhce/http
/var/log/ava/node/chain/2kKSkL4vgVR6QSknoKprsGHVHiEbf7WiGEeSuJUGoB3WzeBhce/http/0.log
/var/log/ava/node/chain/q8CVPqtVEXanDTuU1GnPngLRXVW1WD7XTQi8JVVaJsXgnZDVW
/var/log/ava/node/chain/q8CVPqtVEXanDTuU1GnPngLRXVW1WD7XTQi8JVVaJsXgnZDVW/0.log
/var/log/ava/node/chain/q8CVPqtVEXanDTuU1GnPngLRXVW1WD7XTQi8JVVaJsXgnZDVW/http
/var/log/ava/node/chain/q8CVPqtVEXanDTuU1GnPngLRXVW1WD7XTQi8JVVaJsXgnZDVW/http/0.log
/var/log/ava/node/chain/2mUYSXfLrDtigwbzj1LxKVsHwELghc5sisoXrzJwLqAAQHF4i
/var/log/ava/node/chain/2mUYSXfLrDtigwbzj1LxKVsHwELghc5sisoXrzJwLqAAQHF4i/0.log
/var/log/ava/node/chain/2mUYSXfLrDtigwbzj1LxKVsHwELghc5sisoXrzJwLqAAQHF4i/http
/var/log/ava/node/chain/2mUYSXfLrDtigwbzj1LxKVsHwELghc5sisoXrzJwLqAAQHF4i/http/0.log
/var/log/ava/node/chain/4ktRjsAKxgMr2aEzv9SWmrU7Xk5FniHUrVCX4P1TZSfTLZWFM
/var/log/ava/node/chain/4ktRjsAKxgMr2aEzv9SWmrU7Xk5FniHUrVCX4P1TZSfTLZWFM/0.log
/var/log/ava/node/chain/4ktRjsAKxgMr2aEzv9SWmrU7Xk5FniHUrVCX4P1TZSfTLZWFM/http
/var/log/ava/node/chain/4ktRjsAKxgMr2aEzv9SWmrU7Xk5FniHUrVCX4P1TZSfTLZWFM/http/0.log
/var/log/ava/node/chain/11111111111111111111111111111111LpoYY
/var/log/ava/node/chain/11111111111111111111111111111111LpoYY/0.log
/var/log/ava/node/chain/11111111111111111111111111111111LpoYY/http
/var/log/ava/node/chain/11111111111111111111111111111111LpoYY/http/0.log
/var/log/ava/node/http
/var/log/ava/node/http/0.log
```

danlaine commented:
@moreati Hey, thanks for making this PR. We haven't forgotten it, we're just really busy :sweat_smile:. Thanks for your patience

moreati commented:
No problem, it needs some work for the new staking key location

On Tue, 12 May 2020 at 21:25, danlaine <notifications@github.com> wrote:

> @moreati <https://github.com/moreati> Hey, thanks for making this PR. We
> haven't forgotten it, we're just really busy 😅. Thanks for your patience
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ava-labs/gecko/pull/151#issuecomment-627574052>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABKS4WSZQJDVS52TDYUCS3RRGWDHANCNFSM4M47EQ5A>
> .
>


-- 
Alex Willmer <alex@moreati.org.uk>

moreati commented:
<details>
<summary>Example upgrade</summary>

```
± ansible-playbook -i droplet.yml service_playbook.yml --diff

PLAY [Configure AVA service] ****************************************************************************

TASK [Gathering Facts] **********************************************************************************
ok: [droplet]

TASK [ava-base : Install deps] **************************************************************************
ok: [droplet]

TASK [gopath : Set GOPATH] ******************************************************************************
ok: [droplet]

TASK [ava-build : Update git clone] *********************************************************************
diff --git a/api/keystore/service.go b/api/keystore/service.go
index 7ca34b5..fe1b349 100644
--- a/api/keystore/service.go
+++ b/api/keystore/service.go
@@ -301,6 +301,70 @@ func (ks *Keystore) ImportUser(r *http.Request, args *ImportUserArgs, reply *Imp
 	return nil
 }
 
+// DeleteUserArgs are arguments for passing into DeleteUser requests
+type DeleteUserArgs struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
...
changed: [droplet]

TASK [ava-build : Build project] ************************************************************************
changed: [droplet]

TASK [ava-user : Create AVA daemon group] ***************************************************************
ok: [droplet]

TASK [ava-user : Create AVA daemon user] ****************************************************************
[WARNING]: The value False (type bool) in a string field was converted to 'False' (type string). If this
does not look like what you expect, quote the entire value to ensure it does not change.
ok: [droplet]

TASK [ava-install : Create directories] *****************************************************************
ok: [droplet] => (item=/var/lib/ava)
--- before
+++ after
@@ -1,7 +1,7 @@
 {
-    "group": 0,
-    "mode": "0755",
-    "owner": 0,
+    "group": 996,
+    "mode": "0500",
+    "owner": 996,
     "path": "/var/lib/ava/staking",
-    "state": "absent"
+    "state": "directory"
 }

changed: [droplet] => (item=/var/lib/ava/staking)
ok: [droplet] => (item=/var/log/ava)
ok: [droplet] => (item=/usr/lib/ava/plugins)

TASK [ava-install : Install binary] *********************************************************************
changed: [droplet]

TASK [ava-install : Install plugins] ********************************************************************
changed: [droplet] => (item={'path': '~/go/src/github.com/ava-labs/gecko/build/plugins/evm'})

TASK [ava-upgrade : Check for Gecko 0.2.0 staking key] **************************************************
ok: [droplet]

TASK [ava-upgrade : Check for Gecko newer staking key] **************************************************
ok: [droplet]

TASK [ava-upgrade : Move staking key] *******************************************************************
changed: [droplet]

TASK [ava-upgrade : Check for Gecko 0.2.0 staking certificate] ******************************************
ok: [droplet]

TASK [ava-upgrade : Check for Gecko newer staking certificate] ******************************************
ok: [droplet]

TASK [ava-upgrade : Migrate staking certificate] ********************************************************
changed: [droplet]

TASK [ava-service : Configure AVA service] **************************************************************
--- before: /etc/systemd/system/ava.service
+++ after: /home/alex/.ansible/tmp/ansible-local-13995rodgh9k8/tmp6w9f48ls/ava.service
@@ -17,7 +17,9 @@
             --db-dir="/var/lib/ava/db" \
             --plugin-dir="/usr/lib/ava/plugins" \
             --log-dir="/var/log/ava" \
-            --log-level="info"
+            --log-level="info" \
+            --staking-tls-cert-file="/var/lib/ava/staking/staker.crt" \
+            --staking-tls-key-file="/var/lib/ava/staking/staker.key"
 
 [Install]
 WantedBy=multi-user.target

changed: [droplet]

TASK [ava-service : Enable AVA service] *****************************************************************
ok: [droplet]

RUNNING HANDLER [ava-service : Reload systemd] **********************************************************
ok: [droplet]

RUNNING HANDLER [ava-service : Restart AVA service] *****************************************************
changed: [droplet]

PLAY RECAP **********************************************************************************************
droplet                    : ok=20   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
</details>

holisticode commented on scripts/ansible/ansible.cfg:
```diff
@@ -1,6 +1,16 @@
+any_errors_fatal = true
 transport = ssh
 deprecation_warnings = false
+host_key_checking = false
```
Do we really want this at the global config level?

moreati replied:
I'm happy to remove it, but note I haven't changed the setting from what is master (as `ssh_args = ... o StrictHostKeyChecking=no`), merely the method by which it's configured.

holisticode replied:
You're right.

holisticode commented on scripts/ansible/ansible.cfg:
```diff
@@ -1,6 +1,16 @@
+    -o ControlMaster=auto
+    -o ControlPersist=60s
+    -o ForwardAgent=yes
+    -o UserKnownHostsFile=/dev/null
```
I am not sure about this entry either at the global level?

moreati replied:
Again, I haven't added `UserKnownHostsFile=/dev/null`, I just happened to reformat that line.

Perhaps your concern is that this ansible.cfg has any effect outside this directory? It is only in effect when this directory is the working directory.

holisticode commented on scripts/ansible/roles/ava-base/tasks/main.yml:
```diff
@@ -0,0 +1,16 @@
+      - curl
+      - g++
+      - golang-go  # Assumes Ubuntu 20.04, where this installs Go 1.13
+      - libssl-dev
```
Note: These work only on ubuntu.

holisticode commented on scripts/ansible/roles/ava-base/tasks/main.yml:
```diff
@@ -0,0 +1,16 @@
+      - cmake
+      - curl
+      - g++
+      - golang-go  # Assumes Ubuntu 20.04, where this installs Go 1.13
```
Can we safely make this config choice of Ubuntu 20.04 already?

holisticode commented on scripts/ansible/roles/ava-install/defaults/main.yml:
```diff
@@ -0,0 +1,8 @@
+ava_daemon_bin_dir: "/usr/bin"
```
I personally would prefer `/usr/local/bin` - this will be the default installation path for binary releases.

moreati replied:
Fair point, I'll change it.

holisticode commented on scripts/ansible/roles/ava-install/defaults/main.yml:
```diff
@@ -0,0 +1,8 @@
+ava_daemon_bin_dir: "/usr/bin"
+ava_daemon_data_dir: "/var/lib/{{ ava_daemon_user }}"
+ava_daemon_db_dir: "{{ ava_daemon_data_dir }}/db"
```
@StephenButtolph just recently had defined that the data dir would be `~/.gecko`. But maybe for ansible deployments this is fine here?

In any case, note that due to `ava_daemon_user: ava`, this means that only one user will be able to run `ava` on such a deployment.

moreati replied:
I would advocate that having the directory `/var/lib/ava/.gecko/{staking,db}` is not what a user would expect`. Hence why I overrode it.

moreati replied:
It's worth pointing out that `ava_daemon_user: ava` and others are only default values. Ansible is _very_ flexible about overriding variables. For instance one could do

```yaml
- hosts: localhost
  become: true
  roles:
    - {name: ava-user, vars: {ava_daemon_user: snoopy}
    - {name: ava-user, vars: {ava_daemon_user: woodstock}
```

tyler-smith commented:
Hey @moreati thanks for putting this together, it looks great! I had some concerns but most of them appear to have all been addressed. You have "Install under /usr/local" listed as a todo, is this still in progress? Do you have any other changes you're planning on for this PR?

I would really like to split the OS and OS-version (Ubuntu/20.x) requirements out by putting them in their own files and including them conditionally based on the platform, but this could also be done as a follow up PR IMO. @holisticode, I'll let you make the call on that since you're using this right now and I'm not sure if the hard Ubuntu 20.x requirement works for you or not.

holisticode commented:
If there is no general opposition, I myself am in favor of using ubuntu 20.x right away because it ships right away with a `go` version which `gecko` can run with. Otherwise we need manual installation.

@tyler-smith not sure what you mean by conditionally based on platform, the platform is ubuntu and we use these ansible to just deploy our own nodes, so why would we conditionally support 18.x and 20.x instances?

moreati commented:
> You have "Install under /usr/local" listed as a todo, is this still in progress?

Yes, probably this weekend

> Do you have any other changes you're planning on for this PR?

Not that I can think of. It's already large enough.

moreati commented:
> > Do you have any other changes you're planning on for this PR?
>
> Not that I can think of. It's already large enough.

I added support for (and defaulted to) `--http-host="localhost"`

moreati commented:
To remove any doubt: I believe I've addressed all comments, and this is ready for review.

moreati commented:
Thanks, I'll take a look at the other requirements this weekend. I want to add Raspberry Pi OS as well.
